### PR TITLE
Add language server for Terraform

### LINKED ISSUE #5098: Terraform support
### Check for existing issues

- [X] Completed

### Is your feature request related to a problem?

Missing support for Terraform/HCL

### Describe the solution you'd like

Add support for [terraform-ls](https://github.com/hashicorp/terraform-ls) so that we can have syntax highlighting and auto-complete for terraform configurations

### If applicable, add mockups / screenshots to help present your vision of the feature

N/a

### PR DESCRIPTION
**Pull request by dbanck**
* Depends on: https://github.com/zed-industries/zed/pull/7449
* Closes: https://github.com/zed-industries/zed/issues/5098

---

This PR adds support for downloading and running the Terraform language server for `*.tf` and `*.tfvars` files.  I've verified that the code works for `aarch64` and `x86_64` macOS. Downloading new language server versions on release also works as expected.

Furthermore this PR adds:
- A short docs page for Terraform
- An icon for `*.tf` and `*.tfvars` files

### File Icons

CleanShot 2024-02-10 at 23 10 13@2x

### Completion

CleanShot 2024-02-13 at 20 54 15@2x

### Hover

CleanShot 2024-02-13 at 20 53 40@2x

### Formatting

and more!

## Known issue(s)

@fdionisi discovered that sometimes completion results are inserted with the wrong indentation. Or rather, if you look closely, they are inserted with the correct indentation and then something shifts the closing `}`. I don't think this is related to LSP support and can be addressed in a separate PR.

Release Notes:

- Add language server support for Terraform ([#5098](https://github.com/zed-industries/zed/issues/5098)).

### CODE REVIEW

File: crates/zed/src/languages/terraform.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            let unzip_status = smol::process::Command::new("unzip")
+                .current_dir(&container_dir)
+                .arg(&zip_path)
+                .arg("-d")
+                .arg(&version_dir)
+                .output()
+                .await?
+                .status;
+            if !unzip_status.success() {
+                Err(anyhow!("failed to unzip deno archive"))?;
```
Comment (niklasmtj) [inline crates/zed/src/languages/terraform.rs:88]: This seems to not been updated ☺️ 
```suggestion
                Err(anyhow!("failed to unzip terraform archive"))?;
```

Comment (dbanck) [reply crates/zed/src/languages/terraform.rs]: Good catch!

**Review by riosje — APPROVED** [review]
go for it, release this.

**Comment by yessen-chinapiel-cko** [comment]
heya when is this getting merged ?

**Comment by samuelbailey123** [comment]
Any timeline on release?

**Comment by dbanck** [comment]
I have updated the PR to reflect the new language structure introduced in https://github.com/zed-industries/zed/pull/8270 to resolve the merge conflict.

Let me know if there is anything else I should address.

**Review by nwalke — APPROVED** [review]

**Comment by refl3ction** [comment]
Hello @dbanck! I don't know if it's just me, but I have an issue with the Terraform Language Server when I use Hashcorp's Terraform extension on vscode. For each instance of VSCode I'm running with the extension enabled, it starts a new instance of the `terraform-ls` and each instance of the ls uses on avg 1GB RAM. How is this gonna work for Zed? Once I open any Terraform file it will download/start `terraform-ls`? I'm sorry but I'm not really sure how this will work under the hood 😅

**Comment by maxbrunsfeld** [comment]
Nice work @dbanck. This will ship in Zed 0.125.

**Comment by dbanck** [comment]
@refl3ction It's common for editors to treat a new instance of the editor as a separate thing. I did a quick test with Zed and it behaves similarly to VS Code, spawning a new language server per window. Our language server supports multi-folder workspaces ([VS Code Docs](https://code.visualstudio.com/docs/editor/workspaces), Zed also supports this). If you add more than one folder to a workspace, a single terraform-ls process will be used.

Feel free to file an issue at [vscode-terraform](https://github.com/hashicorp/vscode-terraform/) or [terraform-ls](https://github.com/hashicorp/terraform-ls/) if you think we can improve this. The memory usage is high, but can be common on large workspaces with many modules. It's an ongoing process for us to improve this.

**Pull request merged**